### DISCUSSION TITLE
Dynamic tasks with DatabricksSubmitRunOperator assigns the wrong `spark_env_vars`

### DISCUSSION THREAD

**Discussion by germanium** [apache/airflow#34463] [General] [answered]
### Apache Airflow version

2.7.1

### What happened

Passing environment variables dynamically to the `DatabricksSubmitRunOperator` results in all tasks having the environment variable from the last task in the list 

```python
from airflow.providers.databricks.operators.databricks import DatabricksSubmitRunOperator
from airflow.models import DAG
from datetime import datetime

DAG_INSTANCE = DAG(
    dag_id="my_test_dag",
    start_date=datetime(year=2020, month=11, day=23),
)

env_vars = {
    "CONST_VAR": "some_var"
}

for my_var in ["var_a", "var_b"]:
    env_vars["MY_VAR"] = my_var

    DatabricksSubmitRunOperator(
        task_id=f"processing_{my_var}",
        dag=DAG_INSTANCE,
        retries=1,
        json={
            "run_name": f"processing_{my_var}",
            "new_cluster": {
                "spark_version": "13.1.x-scala2.12"
                "spark_env_vars": env_vars,
            }
        }
    )
```

The UI shows that `MY_VAR` was assigned `var_b`, the last variable in the list for the two dynamic tasks.

Task Instance details for `processing_var_a`
```
{
    "new_cluster": {
        "spark_env_vars": {
            "MY_VAR": "var_b",
            "CONST_VAR": "some_var"
        },
        "spark_version": "13.1.x-scala2.12"
    },
    "run_name": "processing_var_a"
}
```

Task Instance details for `processing_var_b`
```
{
    "new_cluster": {
        "spark_env_vars": {
            "MY_VAR": "var_b",
            "CONST_VAR": "some_var"
        },
        "spark_version": "13.1.x-scala2.12"
    },
    "run_name": "processing_var_b"
}
```

### What you think should happen instead

Each task should have `spark_env_vars` dynamically assigned : 

Task `processing_var_a` should have `MY_VAR=var_a`, task `processing_var_b` `MY_VAR=var_b`

The correct behavior was observed up until `apache-airflow-providers-databricks = "4.0.0"`

### How to reproduce

Create a DAG similar to the one in the `What happened` section and run it.

### Operating System

debian:11-slim

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon = "3.4.0"
apache-airflow-providers-cncf-kubernetes = "4.0.2"
apache-airflow-providers-databricks = "4.5.0"
apache-airflow-providers-datadog = "2.0.4"
apache-airflow-providers-mongo = "2.3.3"
apache-airflow-providers-pagerduty = "2.1.3"
apache-airflow-providers-postgres = "4.1.0"
apache-airflow-providers-slack = "4.2.3"
apache-airflow-providers-snowflake = "2.7.0"

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by Taragolis** [comment, marked as answer]
This is expected, dictionary in Python is a mutable type

```python
d1 = {"a": 1}

d2 = d1
assert d2 == d1
assert d2 is d1

d3 = d1.copy()
assert d3 == d2 == d1
assert d3 is not d1
assert d3 is not d2

d3["a"] = 3.14
assert d2["a"] == 1
assert d1["a"] == 1
assert d3 != d2
assert d3 != d1

d2["a"] = 42
assert d1["a"] == 1, d1
# AssertionError: {'a': 42}
```

So you need to manually make a shallow or a deep copy of dictionary before changes and assign new object, for example

```python

env_vars = {
    "CONST_VAR": "some_var"
}

for my_var in ["var_a", "var_b"]:
    env_vars_copy =  env_vars.copy()
    env_vars_copy["MY_VAR"] = my_var

    DatabricksSubmitRunOperator(
        task_id=f"processing_{my_var}",
        dag=DAG_INSTANCE,
        retries=1,
        json={
            "run_name": f"processing_{my_var}",
            "new_cluster": {
                "spark_version": "13.1.x-scala2.12"
                "spark_env_vars": env_vars_copy,
            }
        }
    )

```

**Reply by germanium** [reply]
That's it! Thanks for your help. It threw me off that this worked up to apache-airflow-providers-databricks 4.0.0.

**Reply by Taragolis** [reply]
> It threw me off that this worked up to apache-airflow-providers-databricks 4.0.0.

It work because internally DatabricksSubmitRunOperator before created copy of the object during assignment to the object attribute.

https://github.com/apache/airflow/blob/12c3c39d1a816c99c626fe4c650e88cf7b1cc1bc/airflow/providers/databricks/operators/databricks.py#L355-L356

https://github.com/apache/airflow/blob/12c3c39d1a816c99c626fe4c650e88cf7b1cc1bc/airflow/providers/databricks/utils/databricks.py#L49-L50

Previous implementation also have side effect if doesn't provide new object. And it is not Airflow specific, it is how it works in Python.

```python
d1 = {"foo": "bar"}

for x in ["a", "b", "c"]:
    if x == "b":
        d1["spam"] = "egg"
    d1["my_var"] = x

    d1_copy = d1.copy()  # Emulate copy during assignments
    print(f"{x!r} without copy dict {d1_copy}")

#  'a' without copy dict {'foo': 'bar', 'my_var': 'a'}
#  'b' without copy dict {'foo': 'bar', 'my_var': 'b', 'spam': 'egg'}
#  'c' without copy dict {'foo': 'bar', 'my_var': 'c', 'spam': 'egg'}

d2 = {"foo": "bar"}

for x in ["a", "b", "c"]:
    d2_copy = d2.copy()
    if x == "b":
        d2_copy["spam"] = "egg"
    d2_copy["my_var"] = x

    d2_copy_of_copy = d2.copy()  # Emulate copy during assignments
    print(f"{x!r} with shallow copy dict {d2_copy}")

# 'a' with shallow copy dict {'foo': 'bar', 'my_var': 'a'}
# 'b' with shallow copy dict {'foo': 'bar', 'spam': 'egg', 'my_var': 'b'}
# 'c' with shallow copy dict {'foo': 'bar', 'my_var': 'c'}
```
